Improve configuration validation

onigoetz opened the pull request:
Fixes #567 

This pull request introduces a custom validator that replaces `jest-validate`

Improvements over the previous implementations are: 
- You can now set a function as a renderer
- Validations can be more precise
- Linters are validated (check that key is a glob, and that values are a string or array of strings)
- valid values for `renderer` are now explicit, where any string was possible before
- Check that `ignore` is an array of strings

I tried to use `joy` `yup` and `schema-utils` to express the schema to validate, but it felt more verbose and less flexible with regard to the number of possible options.

I just ensured that current tests pass, I can write more tests if needed.

okonet commented:
I’m wondering why you decided to ditch the jest-validate since it was giving the flexibility of providing it with updated config a and get the deprecation warnings automatically. Also lots of these now hard coded checks does make the code less maintable. Have you considered adding the fix as a special case like we did with globs?

onigoetz commented:
Because in the current version, jest-validate can't handle different types in the same fields, this will come in version 24 (https://github.com/facebook/jest/blob/master/packages/jest-validate/src/index.js#L18)

Also, jest-validate while being easier to setup is way less precise as it stops at the type of value in the field.

onigoetz commented:
I'll check for alternative solutions then

onigoetz commented:
Hi @okonet, I changed the implementation to use `yup` instead. 
The code is smaller in terms of boilerplate, but there is still more code than before.

In exchange, the configuration validation is more precise and I added more tests to ensure it won't take you by surprise in the future.

okonet commented:
BTW I noticed that Jest v24 was released but I guess jest-validate still has some limitations, right?

okonet commented:
:tada: This PR is included in version 8.1.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/lint-staged)
- [GitHub release](https://github.com/okonet/lint-staged/releases/tag/v8.1.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

onigoetz commented:
Awesome, thanks for the release :)

I also saw it was released, It solves the issue of having multiple types for the same field, which is one part of the PR. It doesn't help have more precise diagnostics on the types and linter object
